oak-jcr/src/test/java/org/apache/jackrabbit/oak/jcr/ImportOperationsTest.java

```diff
@@ -0,0 +1,222 @@
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.UUID;
+
+import javax.jcr.Node;
+import javax.jcr.Property;
+import javax.jcr.PropertyIterator;
+import javax.jcr.PropertyType;
+import javax.jcr.ReferentialIntegrityException;
+import javax.jcr.RepositoryException;
+import javax.jcr.Session;
+import javax.jcr.nodetype.ConstraintViolationException;
+import javax.jcr.nodetype.NodeType;
+
+import org.apache.jackrabbit.oak.fixture.NodeStoreFixture;
+import org.junit.Before;
+import org.junit.Test;
+
+/**
+ * Test coverage for common JCR operations related to importing content.
+ * <p>
+ * Note that the purpose of these tests is not to check conformance with the JCR
+ * specification, but to observe the actual behavior of the implementation
+ * (which may be hard to change).
+ */
+public class ImportOperationsTest extends AbstractRepositoryTest {
+
+    private Session session;
+    private Node testNode;
+    private static String TEST_NODE_NAME = "ImportOperationsTest";
+    private static String TEST_NODE_NAME_REF = "ImportOperationsTest-Reference";
+    private static String TEST_NODE_NAME_TMP = "ImportOperationsTest-Temp";
+
+    public ImportOperationsTest(NodeStoreFixture fixture) {
+        super(fixture);
+    }
+
+    @Before
+    public void setup() throws RepositoryException {
+        this.session = getAdminSession();
+        this.testNode = session.getRootNode().addNode("import-tests", NodeType.NT_UNSTRUCTURED);
+        session.save();
+    }
+
+    @Test
+    public void jcrMixinCreatedOnNtUnstructured() throws RepositoryException {
+        Node test = testNode.addNode(TEST_NODE_NAME, NodeType.NT_UNSTRUCTURED);
+        try {
+            test.setProperty("jcr:created", false);
+            session.save();
+            test.addMixin(NodeType.MIX_CREATED);
+            session.save();
+            // surprise: setting the mixin succeeds, and the subsequent state of
+            // the node is inconsistent with the mixin's type definition
+            assertEquals(false, test.getProperty("jcr:created").getBoolean());
+        } finally {
+            test.remove();
+            session.save();
+        }
+    }
+
+    @Test
+    public void jcrMixinReferenceableOnNtUnstructuredBeforeSettingMixin() throws RepositoryException {
+        Node test = testNode.addNode(TEST_NODE_NAME, NodeType.NT_UNSTRUCTURED);
+        try {
+            String testUuid = UUID.randomUUID().toString();
+            test.setProperty("jcr:uuid", testUuid);
+            session.save();
+            test.addMixin(NodeType.MIX_REFERENCEABLE);
+            session.save();
+            // surprise: setting the mixin succeeds, and the subsequent state of
+            // the node doesn't have an autocreated UUID
+            assertEquals(testUuid, test.getProperty("jcr:uuid").getString());
```

Comment:
why should it get a new id assigned if the previous one does not have any conflicts? I think a test where you assign an already taken value and later on add the mixin is more interesting.

Reply:
well, the spec requires the jcr:uuid to be "autocreated", which IMHO does not mean "supplied by the caller". I *do* agree that another test would be helpful here.

Reply:
Done (and confirmed that this fails with exception, altough it happes early - before mixin:ref was added).